_posts/2020-09-15-mutiny-pagination.adoc

```diff
@@ -0,0 +1,200 @@
+So, to retrieve all the beers, we need to invoke the service for page 1, 2, 3... until the returned JSON array is empty. 
+
+In an imperative word, to retrieve all the beers, you would do something like this:
+
+[source, java]
+----
+List<Beer> beers = ...;
+int page = 1;
+List<Beer> batch = ...
+do {
+  batch= getBeersFromPage(page);
+  beers.addAll(batch);
+  page = page + 1;
+} while (! batch.isEmpty());
+----
+
+How can we achieve the same in a reactive manner and build a stream of beer?
+
+image:/assets/images/posts/mutiny-pagination/mutiny-pagination.png[]
+
+Let's proceed step by step.
+
+## Retrieving a single page
+
+First, we need to see how we can retrieve a single page.
+I'm going to use the Vert.x Web Client, but you can use any reactive HTTP clients providing a Mutiny API.
+
+[source, java]
+----
+// Create the client
+WebClient client = WebClient.create(vertx, new WebClientOptions()
+      .setDefaultHost("api.punkapi.com")
+      .setDefaultPort(443)
+      .setSsl(true)
+);
+
+// Retrieve the first page
+Uni<List<Beer>> uni = client.get("/v2/beers?page=1")
+      .send()
+      .onItem().transform(Pagination::toListOfBeer);
+----
+
+This snippet creates the web client. 
+Then, we use that client and retrieve the first page.
+
+When we receive the result (`onItem`), we transform the JSON array into a list of beers.
+
+Let's extract this code in a method and take the page number as parameter:
+
+[source, java]
+----
+private static Uni<List<Beer>> getPage(WebClient client, int page) {
+    return client.get("/v2/beers?page=" + page)
+            .send()
+            .onItem().transform(Pagination::toListOfBeer);
+}
+----
+
+So far, so good.
+
+## Retrieving multiple page
+
+So, now, we know how to retrieve a single page and extract the items from it. 
+We just need to repeat this operation for every page, and provide a stream.
+
+Mutiny provides a method to create a `Multi` by repeating multiple times a `Uni`.
+Under the hood, it calls a method returning a `Uni` and subscribe on it.
+But we need to make _progress_, and pass the current page.
+Mutiny offers the possibility to store a state, to the method creating the `Uni` can increment and get the page number:
+
+[source, java]
+----
+Multi.createBy().repeating().uni(AtomicInteger::new, page ->
+		getPage(client, page.incrementAndGet())
+)
+----
+
+This code above creates a stream with the item emitted by the `Unis` returned by the `getPage` method.
+We increment the page number (stored in an `AtomicInteger`) every time.
+So, it retrieves the page 1, 2, 3 ... and every time emits the received `List<Beer>` downstream.
+
+However, at some point, we must stop. 
+As we said earlier, we can stop when the returned list is empty:
+
+[source, java]
+----
+Multi<List<Beer>> multi = Multi.createBy().repeating().uni(AtomicInteger::new, page ->
+     getPage(client, page.incrementAndGet())
+)
+.until(List::isEmpty);
```

Comment:
For my own understanding, is the `until` waiting for the `List<Beer>` on the `Mutli` to be empty, or the `List` returned from `getPage`?

Reply:
the list from the `getPage()` method, which means that it waits until the received item (the list) is empty.

Reply:
Ok, I've added a bit more explanation, can you check?

Reply:
is that because the `until` is essentially part of the `repeating`?

Reply:
yes, it is. `repeating` returns an object to configure when the repetition stops, not a multi. `until` returns the `multi`.